docs/technical-design.md

```diff
@@ -0,0 +1,130 @@
+# Technical Design
+
+## To be discussed
+
+
+## Istio operator
+
+![IstioOperator Overview](./istio-operator-overview.svg)
+
+### Ownership of current resources in Kyma repository
+
+In the transitioning to a more modularised architecture, the [IstioOperator resource](https://github.com/kyma-project/kyma/tree/main/resources/istio), 
+the [additional istio-resources](https://github.com/kyma-project/kyma/tree/main/resources/istio-resources) and 
+the [certificates](https://github.com/kyma-project/kyma/tree/main/resources/certificates) must be moved to the new modules.
+
+#### IstioOperator resource
+This is moved into our new Istio operator and used to define the default values for Istio, which can be customised by the user through `IstioCR`.
+
+#### istio-resources
+
+##### Istio Grafana dashboards
+It's still needs to be decided who will have the ownership of the dashboards. There are different things to consider like the change interval or relevance of Istio version updates.
+
+##### Istio ServiceMonitor & istio-healthz Virtual Service
+They provide monitoring capability for the Istio installation and should therefore be reconciled by the operator.
+
+##### Global mTLS PeerAuthentication
+This is tightly coupled to our Istio installation and should therefore be reconciled by the operator.
+
+##### Kyma Gateway
+This should be moved to the API Gateway as it's a default gateway we want to provide, and it has a stronger relation to the responsibilities
+of the API Gateway that to Istio. Since API Gateway is already dependent on Istio, we do not add any additional dependency by moving it. 
+
+#### Certificate resources
+They should also be moved to API Gateway as they are tightly coupled with the `Kyma Gateway` resource.
+
+### Handling of Istio version
+We don't want to expose the version. That means that the version of Istio is coupled to the version of the operator. The benefit of this is,
+that we are in full control of the versioning and hide this complexity from the user.  
+That means if we want to release a new Istio version, we have to release a new version of the operator.
+
+## IstioController
+
+![Component Diagram](./controller-component-diagram.svg)
+
+### Installation & Upgrade of Istio
+The Istio installation, upgrade and uninstall is done using [Istio Go module](https://github.com/istio/istio).
+
+### Reconciliation
+The reconciliation is done by multiple self-contained sub-reconcilers to have a better extensibility and maintainability. This means each of this reconcilers must have its clearly separated responsibility
```

Comment:
I simplified the structure in the component diagram and added some [additional docs](https://github.com/triffer/istio/blob/7ba855023208378cd612d03a9d5782df901dcb7e/docs/technical-design.md?plain=1#L50) to explain our relation between the controller and reconcilers. This means your assumptions that what we called reconcilers are actually controllers doesn't match what we had in mind when creating this modular controller design.
I agree that it might make sense to split our "sub-reconciler" up in multiple controllers, but I see some drawbacks about this. 
Some of the new controllers will use the same CR and rely on fields that are rolled out by the IstioController (e.g. a controller handling the logic of `IstioIngressGatewayReconciler`. Other controllers will be fully time-based and rely on the state of Istio installation, e.g. controller handling the logic of `PeerAuthenticationReconciler`.
Decoupling this from the Istio reconciliation will make the implementation of this controllers more complex and doesn't provide any real benefit in my opinion. Since the logic is already divided into components, it should be easy to move them to separate controllers if needed.

Reply:
I added a simple sequence diagram for the reconciliation loop of `IstioController`. You can find it [here](https://github.com/triffer/istio/blob/technical-design/docs/reconciliation-sequence-diagram.svg). 
We will also rename components that had the suffix `Reconciler` to `Reconciliation`.
I hope this helps to better understand the components.

Reply:
Ok. It is more clear now. Still it is important to avoid sequences and waiting in the reconciliation logic. If you need to invoke second reconciliation (e.g. proxy sidecar) you should not do it in the same main loop invocation as IstioInstallation. You trigger IstioInstallation in the first invocation and you should return RequeAfter. The second invocation should check if installation is completed and execute proxy sidecar reconciliation. I think you know it, but initial description and sequence diagram can be misleading.

Reply:
Reconcile of a controller is meant to be fast and non-waiting process. Installation is actually not really optimal for such a task. I did express this concern a while back, but somehow did not trigger discussion about it. What do you think of moving it out of the reconcile() processing? For instance into a separate Job task? We used before Argo Workflows for various long running tasks. I would like that the Istio operator only control operations of others than to actually do time consuming tasks.

Reply:
Installation can be done this way as it is just about creating some kubernetes resources (request to api server) and waiting for the status. You don't need to wait. You should check the status and if it is not ready exit the reconciliation loop with RequeueAfter. In the next invocation you check it again and if nothing has changed you exit it again. You don't have anything to process (any time consuming task).

Reply:
We install Istio with the CLI which takes time? or I am mistaken something. Current Reconciler run looks like that:
```
{"level":"INFO","time":"2022-12-08T13:36:55.334+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:37:25.348+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:37:38.557+0100","message":"Istio has been updated successfully to version 1.16.0-distroless"}
{"level":"INFO","time":"2022-12-08T13:37:38.557+0100","message":"Restarting ingress-gateway"}
2022-12-08T12:37:42.753965Z	info	klog	Waited for 1.146225584s due to client-side throttling, not priority and fairness, request: PATCH:https://api.vv-test2.goatz.shoot.canary.k8s-hana.ondemand.com/apis/apps/v1/namespaces/vladimir/deployments/httpbin
{"level":"INFO","time":"2022-12-08T13:37:55.350+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:38:25.351+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:38:55.352+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:39:25.352+0100","message":"Component 'istio' has status 'running'"}
{"level":"INFO","time":"2022-12-08T13:39:33.442+0100","message":"Proxy reset for 17 pods successfully done"}
{"level":"WARN","time":"2022-12-08T13:39:34.250+0100","message":"Reconciler Metrics not initialized"}
{"level":"INFO","time":"2022-12-08T13:39:34.395+0100","message":"Component 'istio' has status 'success'"}
```
Looking at the timestamps there is quite something going on. My idea was to extract these time consuming operations outside of the controller. Having them in separate Jobs it may also be easy on analysing logs, when something goes wrong. Not needed to search huge reconciler logs. Let's discuss it tomorrow.

Reply:
Timestamp doesn't say that istio installer is doing anything. Istio uses helm to render manifests, then applies the manifest in the kuybernetes and then waits. And waits. And waits. Active part is probably 1%. The rest is waiting for kubelet to deploy istiod and istioingressgateway.

Reply:
We like the Istio Operator Install, we just are being hesitant because of the `discouraged` note in here:
https://istio.io/latest/docs/setup/install/operator/

Reply:
@pbochynski 
We updated the documentation for the [Istio installation](https://github.com/kyma-project/istio/pull/36/files#diff-60b0177084514f6a20db25c5360a874044da6414583268b928d0056eeceb308eR50). 
We decided to install the operator using the Istio Go module and apply the `IstioOperator CR` to install Istio. This way we avoid the long blocking the reconciliation of our controller during the Istio Installation. We also added a flow diagram to visualise the process of installing, upgrading and uninstalling Istio.

The docs were extended to clarify the scenario when the `IstioVersionManager` (formerly known as `IstioManager`) would be useful.